File: DataConverter.java
// DataConverter.java
import java.io.*;
import java.util.*;
import java.util.regex.*;

class UnitInfo {
    String name;
    double factor;
    UnitInfo(String n, double f) { name = n; factor = f; }
}

public class DataConverter {
    private static final Map<String, UnitInfo> UNITS = new LinkedHashMap<>();
    static {
        UNITS.put("b", new UnitInfo("bit", 1.0/8));
        UNITS.put("B", new UnitInfo("byte", 1));
        UNITS.put("KB", new UnitInfo("kilobyte", 1000));
        UNITS.put("KiB", new UnitInfo("kibibyte", 1024));
        UNITS.put("MB", new UnitInfo("megabyte", 1000*1000));
        UNITS.put("MiB", new UnitInfo("mebibyte", 1024*1024));
        UNITS.put("GB", new UnitInfo("gigabyte", 1000*1000*1000));
        UNITS.put("GiB", new UnitInfo("gibibyte", 1024*1024*1024));
        UNITS.put("TB", new UnitInfo("terabyte", 1000L*1000*1000*1000));
        UNITS.put("TiB", new UnitInfo("tebibyte", 1024L*1024*1024*1024));
        UNITS.put("PB", new UnitInfo("petabyte", 1000L*1000*1000*1000*1000));
        UNITS.put("PiB", new UnitInfo("pebibyte", 1024L*1024*1024*1024*1024));
    }
    private static final List<String> UNIT_ORDER = Arrays.asList("b","B","KB","KiB","MB","MiB","GB","GiB","TB","TiB","PB","PiB");

    private int precision;
    private List<String> history = new ArrayList<>();

    public DataConverter(int prec) { precision = prec; }

    private class Parsed {
        double value; String unit;
        Parsed(double v, String u) { value = v; unit = u; }
    }

    private Parsed parseValueUnit(String s) throws Exception {
        s = s.trim();
        Matcher m = Pattern.compile("^([\\d.]+)\\s*([a-zA-Z]+)$").matcher(s);
        if (!m.matches()) {
            try {
                double val = Double.parseDouble(s);
                return new Parsed(val, "B");
            } catch (NumberFormatException e) {
                throw new Exception("Invalid format. Use 'value unit'.");
            }
        }
        double val = Double.parseDouble(m.group(1));
        String unitStr = m.group(2).toUpperCase();
        String unit = null;
        for (String u : UNITS.keySet()) {
            if (u.equalsIgnoreCase(unitStr)) { unit = u; break; }
        }
        if (unit == null && unitStr.endsWith("B") && unitStr.length() > 1) {
            String shortU = unitStr.substring(0, unitStr.length()-1);
            for (String u : UNITS.keySet()) {
                if (u.equalsIgnoreCase(shortU)) { unit = u; break; }
            }
        }
        if (unit == null) throw new Exception("Unknown unit: " + unitStr);
        return new Parsed(val, unit);
    }

    private double toBytes(double value, String unit) { return value * UNITS.get(unit).factor; }
    private double fromBytes(double bytes, String unit) { return bytes / UNITS.get(unit).factor; }

    public double convert(double value, String fromUnit, String toUnit) {
        double bytes = toBytes(value, fromUnit);
        double result = fromBytes(bytes, toUnit);
        history.add(format(value, fromUnit) + " -> " + format(result, toUnit));
        if (history.size() > 20) history.remove(0);
        return result;
    }

    public Parsed autoConvert(double value, String unit) {
        double bytes = toBytes(value, unit);
        String bestUnit = "B";
        double bestVal = bytes;
        for (String u : UNIT_ORDER) {
            double v = bytes / UNITS.get(u).factor;
            if (v >= 1 && v < 1000) {
                bestUnit = u;
                bestVal = v;
                break;
            }
        }
        if (bestUnit.equals("B") && bytes < 1) {
            bestUnit = "b";
            bestVal = bytes / UNITS.get("b").factor;
        }
        return new Parsed(bestVal, bestUnit);
    }

    public String format(double value, String unit) {
        return String.format("%." + precision + "f %s", value, unit);
    }

    public void showHistory() {
        if (history.isEmpty()) {
            System.out.println("No conversions yet.");
        } else {
            System.out.println("\n--- Conversion History (last 20) ---");
            for (int i = 0; i < history.size(); i++) {
                System.out.println((i+1) + ". " + history.get(i));
            }
        }
    }

    public void printHelp() {
        System.out.println("\nSupported units:");
        for (Map.Entry<String, UnitInfo> e : UNITS.entrySet()) {
            System.out.printf("  %4s = %s (factor %.0f)\n", e.getKey(), e.getValue().name, e.getValue().factor);
        }
    }

    public static void main(String[] args) throws Exception {
        Scanner scanner = new Scanner(System.in);
        DataConverter converter = new DataConverter(2);
        System.out.println("=== Data Unit Converter ===");
        while (true) {
            System.out.println("\n1. Convert single value");
            System.out.println("2. Batch convert from file");
            System.out.println("3. Show conversion history");
            System.out.println("4. Set precision (current: " + converter.precision + ")");
            System.out.println("5. Help / unit info");
            System.out.println("6. Exit");
            System.out.print("Choose: ");
            String choice = scanner.nextLine().trim();
            switch (choice) {
                case "1":
                    try {
                        System.out.print("Enter value and unit (e.g., 1024 MB): ");
                        String inp = scanner.nextLine();
                        Parsed p = converter.parseValueUnit(inp);
                        System.out.print("Target unit (leave blank for auto): ");
                        String target = scanner.nextLine().trim();
                        if (!target.isEmpty()) {
                            String toUnit = null;
                            for (String u : UNITS.keySet()) {
                                if (u.equalsIgnoreCase(target)) { toUnit = u; break; }
                            }
                            if (toUnit == null) {
                                System.out.println("Unknown target unit.");
                                break;
                            }
                            double result = converter.convert(p.value, p.unit, toUnit);
                            System.out.println("\nResult: " + converter.format(p.value, p.unit) + " = " + converter.format(result, toUnit));
                        } else {
                            Parsed auto = converter.autoConvert(p.value, p.unit);
                            converter.history.add(converter.format(p.value, p.unit) + " -> " + converter.format(auto.value, auto.unit));
                            if (converter.history.size() > 20) converter.history.remove(0);
                            System.out.println("\nResult: " + converter.format(p.value, p.unit) + " = " + converter.format(auto.value, auto.unit));
                        }
                    } catch (Exception e) {
                        System.out.println("Error: " + e.getMessage());
                    }
                    break;
                case "2":
                    System.out.print("Enter batch file path: ");
                    String fname = scanner.nextLine().trim();
                    File file = new File(fname);
                    if (!file.exists()) {
                        System.out.println("File not found.");
                        break;
                    }
                    System.out.print("Target unit for all conversions (leave blank for auto): ");
                    String target2 = scanner.nextLine().trim();
                    String toUnit2 = null;
                    if (!target2.isEmpty()) {
                        for (String u : UNITS.keySet()) {
                            if (u.equalsIgnoreCase(target2)) { toUnit2 = u; break; }
                        }
                        if (toUnit2 == null) {
                            System.out.println("Unknown target unit.");
                            break;
                        }
                    }
                    System.out.println("\nBatch results:");
                    try (BufferedReader br = new BufferedReader(new FileReader(file))) {
                        String line;
                        while ((line = br.readLine()) != null) {
                            line = line.trim();
                            if (line.isEmpty()) continue;
                            try {
                                Parsed p2 = converter.parseValueUnit(line);
                                if (toUnit2 != null) {
                                    double result = converter.convert(p2.value, p2.unit, toUnit2);
                                    System.out.println(converter.format(p2.value, p2.unit) + " -> " + converter.format(result, toUnit2));
                                } else {
                                    Parsed auto2 = converter.autoConvert(p2.value, p2.unit);
                                    System.out.println(converter.format(p2.value, p2.unit) + " -> " + converter.format(auto2.value, auto2.unit));
                                }
                            } catch (Exception e) {
                                System.out.println("Skipping '" + line + "': " + e.getMessage());
                            }
                        }
                    }
                    break;
                case "3":
                    converter.showHistory();
                    break;
                case "4":
                    System.out.print("Enter number of decimal places: ");
                    try {
                        int prec = Integer.parseInt(scanner.nextLine().trim());
                        if (prec < 0) throw new NumberFormatException();
                        converter.precision = prec;
                        System.out.println("Precision updated.");
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid precision.");
                    }
                    break;
                case "5":
                    converter.printHelp();
                    break;
                case "6":
                    System.out.println("Goodbye!");
                    scanner.close();
                    return;
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }
}

File: DataConverter.cs
// DataConverter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

class UnitInfo
{
    public string Name { get; set; }
    public double Factor { get; set; }
}

class DataConverter
{
    private static readonly Dictionary<string, UnitInfo> Units = new Dictionary<string, UnitInfo>
    {
        {"b", new UnitInfo{Name="bit", Factor=1.0/8}},
        {"B", new UnitInfo{Name="byte", Factor=1}},
        {"KB", new UnitInfo{Name="kilobyte", Factor=1000}},
        {"KiB", new UnitInfo{Name="kibibyte", Factor=1024}},
        {"MB", new UnitInfo{Name="megabyte", Factor=1000*1000}},
        {"MiB", new UnitInfo{Name="mebibyte", Factor=1024*1024}},
        {"GB", new UnitInfo{Name="gigabyte", Factor=1000*1000*1000}},
        {"GiB", new UnitInfo{Name="gibibyte", Factor=1024*1024*1024}},
        {"TB", new UnitInfo{Name="terabyte", Factor=1000L*1000*1000*1000}},
        {"TiB", new UnitInfo{Name="tebibyte", Factor=1024L*1024*1024*1024}},
        {"PB", new UnitInfo{Name="petabyte", Factor=1000L*1000*1000*1000*1000}},
        {"PiB", new UnitInfo{Name="pebibyte", Factor=1024L*1024*1024*1024*1024}}
    };

    private int precision;
    private List<string> history = new List<string>();

    public DataConverter(int prec = 2) { precision = prec; }

    private (double value, string unit) ParseValueUnit(string s)
    {
        s = s.Trim();
        var m = Regex.Match(s, @"^([\d.]+)\s*([a-zA-Z]+)$");
        if (!m.Success)
        {
            // try just number
            if (double.TryParse(s, out double val))
                return (val, "B");
            throw new Exception("Invalid format. Use 'value unit'.");
        }
        double val2 = double.Parse(m.Groups[1].Value);
        string unitStr = m.Groups[2].Value.ToUpper();
        // normalize
        foreach (var u in Units.Keys)
        {
            if (string.Equals(u, unitStr, StringComparison.OrdinalIgnoreCase))
                return (val2, u);
        }
        // try alias: remove trailing 'B'
        if (unitStr.EndsWith("B") && unitStr.Length > 1)
        {
            string shortU = unitStr.Substring(0, unitStr.Length - 1);
            foreach (var u in Units.Keys)
            {
                if (string.Equals(u, shortU, StringComparison.OrdinalIgnoreCase))
                    return (val2, u);
            }
        }
        throw new Exception($"Unknown unit: {unitStr}");
    }

    private double ToBytes(double value, string unit) => value * Units[unit].Factor;
    private double FromBytes(double bytes, string unit) => bytes / Units[unit].Factor;

    public double Convert(double value, string fromUnit, string toUnit)
    {
        double bytes = ToBytes(value, fromUnit);
        double result = FromBytes(bytes, toUnit);
        history.Add($"{Format(value, fromUnit)} -> {Format(result, toUnit)}");
        if (history.Count > 20) history.RemoveAt(0);
        return result;
    }

    public (double value, string unit) AutoConvert(double value, string unit)
    {
        double bytes = ToBytes(value, unit);
        string[] order = { "b", "B", "KB", "KiB", "MB", "MiB", "GB", "GiB", "TB", "TiB", "PB", "PiB" };
        string bestUnit = "B";
        double bestVal = bytes;
        foreach (var u in order)
        {
            double v = bytes / Units[u].Factor;
            if (v >= 1 && v < 1000)
            {
                bestUnit = u;
                bestVal = v;
                break;
            }
        }
        if (bestUnit == "B" && bytes < 1)
        {
            bestUnit = "b";
            bestVal = bytes / Units["b"].Factor;
        }
        return (bestVal, bestUnit);
    }

    public string Format(double value, string unit) =>
        $"{value.ToString($"F{precision}")} {unit}";

    public void ShowHistory()
    {
        if (history.Count == 0)
        {
            Console.WriteLine("No conversions yet.");
            return;
        }
        Console.WriteLine("\n--- Conversion History (last 20) ---");
        for (int i = 0; i < history.Count; i++)
            Console.WriteLine($"{i+1}. {history[i]}");
    }

    public void PrintHelp()
    {
        Console.WriteLine("\nSupported units:");
        foreach (var kv in Units.OrderBy(kv => kv.Key))
            Console.WriteLine($"  {kv.Key,4} = {kv.Value.Name} (factor {kv.Value.Factor})");
    }

    static void Main()
    {
        var converter = new DataConverter(2);
        Console.WriteLine("=== Data Unit Converter ===");
        while (true)
        {
            Console.WriteLine("\n1. Convert single value");
            Console.WriteLine("2. Batch convert from file");
            Console.WriteLine("3. Show conversion history");
            Console.WriteLine($"4. Set precision (current: {converter.precision})");
            Console.WriteLine("5. Help / unit info");
            Console.WriteLine("6. Exit");
            Console.Write("Choose: ");
            string choice = Console.ReadLine()?.Trim() ?? "";
            switch (choice)
            {
                case "1":
                    try
                    {
                        Console.Write("Enter value and unit (e.g., 1024 MB): ");
                        string inp = Console.ReadLine();
                        var (val, fromUnit) = converter.ParseValueUnit(inp);
                        Console.Write("Target unit (leave blank for auto): ");
                        string target = Console.ReadLine()?.Trim() ?? "";
                        if (!string.IsNullOrEmpty(target))
                        {
                            string toUnit = null;
                            foreach (var u in Units.Keys)
                                if (string.Equals(u, target, StringComparison.OrdinalIgnoreCase))
                                { toUnit = u; break; }
                            if (toUnit == null)
                            {
                                Console.WriteLine("Unknown target unit.");
                                break;
                            }
                            double result = converter.Convert(val, fromUnit, toUnit);
                            Console.WriteLine($"\nResult: {converter.Format(val, fromUnit)} = {converter.Format(result, toUnit)}");
                        }
                        else
                        {
                            var (result, bestUnit) = converter.AutoConvert(val, fromUnit);
                            converter.history.Add($"{converter.Format(val, fromUnit)} -> {converter.Format(result, bestUnit)}");
                            if (converter.history.Count > 20) converter.history.RemoveAt(0);
                            Console.WriteLine($"\nResult: {converter.Format(val, fromUnit)} = {converter.Format(result, bestUnit)}");
                        }
                    }
                    catch (Exception e) { Console.WriteLine($"Error: {e.Message}"); }
                    break;
                case "2":
                    Console.Write("Enter batch file path: ");
                    string fname = Console.ReadLine()?.Trim() ?? "";
                    if (!File.Exists(fname))
                    {
                        Console.WriteLine("File not found.");
                        break;
                    }
                    Console.Write("Target unit for all conversions (leave blank for auto): ");
                    string target2 = Console.ReadLine()?.Trim() ?? "";
                    string toUnit2 = null;
                    if (!string.IsNullOrEmpty(target2))
                    {
                        foreach (var u in Units.Keys)
                            if (string.Equals(u, target2, StringComparison.OrdinalIgnoreCase))
                            { toUnit2 = u; break; }
                        if (toUnit2 == null)
                        {
                            Console.WriteLine("Unknown target unit.");
                            break;
                        }
                    }
                    Console.WriteLine("\nBatch results:");
                    foreach (var line in File.ReadLines(fname))
                    {
                        string l = line.Trim();
                        if (string.IsNullOrEmpty(l)) continue;
                        try
                        {
                            var (val, fromUnit) = converter.ParseValueUnit(l);
                            if (toUnit2 != null)
                            {
                                double result = converter.Convert(val, fromUnit, toUnit2);
                                Console.WriteLine($"{converter.Format(val, fromUnit)} -> {converter.Format(result, toUnit2)}");
                            }
                            else
                            {
                                var (result, bestUnit) = converter.AutoConvert(val, fromUnit);
                                Console.WriteLine($"{converter.Format(val, fromUnit)} -> {converter.Format(result, bestUnit)}");
                            }
                        }
                        catch (Exception e) { Console.WriteLine($"Skipping '{l}': {e.Message}"); }
                    }
                    break;
                case "3":
                    converter.ShowHistory();
                    break;
                case "4":
                    Console.Write("Enter number of decimal places: ");
                    if (int.TryParse(Console.ReadLine(), out int prec) && prec >= 0)
                    {
                        converter.precision = prec;
                        Console.WriteLine("Precision updated.");
                    }
                    else Console.WriteLine("Invalid precision.");
                    break;
                case "5":
                    converter.PrintHelp();
                    break;
                case "6":
                    Console.WriteLine("Goodbye!");
                    return;
                default:
                    Console.WriteLine("Invalid choice.");
                    break;
            }
        }
    }
}
